File: repositories/post-repository.go
package repositories

import (
	"context"
	"regexp"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"tech-letter/models"
)

type PostRepository struct {
	col *mongo.Collection
}

func NewPostRepository(db *mongo.Database) *PostRepository {
	return &PostRepository{col: db.Collection("posts")}
}

// IsExistByLink checks if a post exists by its link.
func (r *PostRepository) IsExistByLink(ctx context.Context, link string) (bool, error) {
	err := r.col.FindOne(ctx, bson.M{"link": link}).Err()
	if err == mongo.ErrNoDocuments {
		return false, nil
	}
	return err == nil, err
}

// Insert inserts a new post document.
func (r *PostRepository) Insert(ctx context.Context, p *models.Post) (*mongo.InsertOneResult, error) {
	now := time.Now()
	if p.CreatedAt.IsZero() {
		p.CreatedAt = now
	}
	p.UpdatedAt = now
	if p.Status == (models.StatusFlags{}) {
		p.Status = models.StatusFlags{}
	}
	return r.col.InsertOne(ctx, p)
}

// FindByLink returns a post by link
func (r *PostRepository) FindByLink(ctx context.Context, link string) (*models.Post, error) {
	var p models.Post
	if err := r.col.FindOne(ctx, bson.M{"link": link}).Decode(&p); err != nil {
		return nil, err
	}
	return &p, nil
}

// SetAISummarized updates only the status.ai_summarized flag and updated_at.
// 다른 상태 필드는 건드리지 않으므로, 향후 StatusFlags 필드가 늘어나더라도 안전하다.
func (r *PostRepository) SetAISummarized(ctx context.Context, postID interface{}, summarized bool) error {
	_, err := r.col.UpdateByID(ctx, postID, bson.M{
		"$set": bson.M{
			"status.ai_summarized": summarized,
			"updated_at":           time.Now(),
		},
	})
	return err
}

// UpdateAISummary sets normalized summary snapshot on the post document
func (r *PostRepository) UpdateAISummary(ctx context.Context, postID interface{}, summary models.AISummary) error {
	_, err := r.col.UpdateByID(ctx, postID, bson.M{
		"$set": bson.M{"aisummary": summary, "updated_at": time.Now()},
	})
	return err
}

// UpdateThumbnailURL sets thumbnail_url field
func (r *PostRepository) UpdateThumbnailURL(ctx context.Context, postID interface{}, url string) error {
	_, err := r.col.UpdateByID(ctx, postID, bson.M{
		"$set": bson.M{"thumbnail_url": url, "updated_at": time.Now()},
	})
	return err
}

// SetThumbnailParsed updates only the status.thumbnail_parsed flag and updated_at.
// AISummarized와 마찬가지로, 다른 상태 필드는 건드리지 않는다.
func (r *PostRepository) SetThumbnailParsed(ctx context.Context, postID interface{}, parsed bool) error {
	_, err := r.col.UpdateByID(ctx, postID, bson.M{
		"$set": bson.M{
			"status.thumbnail_parsed": parsed,
			"updated_at":              time.Now(),
		},
	})
	return err
}

type ListPostsOptions struct {
	Page       int
	PageSize   int
	Categories []string
	Tags       []string
	BlogID     *primitive.ObjectID
	BlogName   string
}

// List returns posts with filters and pagination, sorted by published_at desc
func (r *PostRepository) List(ctx context.Context, opt ListPostsOptions) ([]models.Post, int64, error) {
	filter := bson.M{}
	// Build case-insensitive anchored regex arrays for categories and tags
	toRegexIn := func(values []string) []interface{} {
		arr := make([]interface{}, 0, len(values))
		for _, v := range values {
			if v == "" {
				continue
			}
			pattern := "^" + regexp.QuoteMeta(v) + "$"
			arr = append(arr, primitive.Regex{Pattern: pattern, Options: "i"})
		}
		return arr
	}

	catsRegex := toRegexIn(opt.Categories)
	tagsRegex := toRegexIn(opt.Tags)
	if len(catsRegex) > 0 && len(tagsRegex) > 0 {
		filter["$or"] = []bson.M{
			{"aisummary.categories": bson.M{"$in": catsRegex}},
			{"aisummary.tags": bson.M{"$in": tagsRegex}},
		}
	} else if len(catsRegex) > 0 {
		filter["aisummary.categories"] = bson.M{"$in": catsRegex}
	} else if len(tagsRegex) > 0 {
		filter["aisummary.tags"] = bson.M{"$in": tagsRegex}
	}

	// Blog filters
	if opt.BlogID != nil {
		filter["blog_id"] = *opt.BlogID
	}
	if opt.BlogName != "" {
		filter["blog_name"] = primitive.Regex{Pattern: "^" + regexp.QuoteMeta(opt.BlogName) + "$", Options: "i"}
	}

	if opt.Page <= 0 {
		opt.Page = 1
	}
	if opt.PageSize <= 0 || opt.PageSize > 100 {
		opt.PageSize = 20
	}
	skip := int64((opt.Page - 1) * opt.PageSize)
	limit := int64(opt.PageSize)

	total, err := r.col.CountDocuments(ctx, filter)
	if err != nil {
		return nil, 0, err
	}

	findOpts := options.Find().SetSkip(skip).SetLimit(limit).SetSort(bson.D{
		{Key: "published_at", Value: -1},
		{Key: "_id", Value: -1},
	})
	cur, err := r.col.Find(ctx, filter, findOpts)
	if err != nil {
		return nil, 0, err
	}
	defer cur.Close(ctx)

	var results []models.Post
	for cur.Next(ctx) {
		var p models.Post
		if err := cur.Decode(&p); err != nil {
			return nil, 0, err
		}
		results = append(results, p)
	}
	if err := cur.Err(); err != nil {
		return nil, 0, err
	}
	return results, total, nil
}

// FindByID returns a post by its ObjectID
func (r *PostRepository) FindByID(ctx context.Context, id primitive.ObjectID) (*models.Post, error) {
	var p models.Post
	if err := r.col.FindOne(ctx, bson.M{"_id": id}).Decode(&p); err != nil {
		return nil, err
	}
	return &p, nil
}

// IncrementViewCount increments the view_count field by 1 for the given post ID
func (r *PostRepository) IncrementViewCount(ctx context.Context, id primitive.ObjectID) error {
	_, err := r.col.UpdateByID(ctx, id, bson.M{
		"$inc": bson.M{"view_count": 1},
		"$set": bson.M{"updated_at": time.Now()},
	})
	return err
}

// UpdateFields updates specific fields of a post
func (r *PostRepository) UpdateFields(ctx context.Context, id primitive.ObjectID, updates map[string]interface{}) error {
	update := bson.M{
		"$set": bson.M{
			"updated_at": time.Now(),
		},
	}
	for k, v := range updates {
		update["$set"].(bson.M)[k] = v
	}
	_, err := r.col.UpdateByID(ctx, id, update)
	return err
}

// FindPostsWithoutRenderedHTML RenderedHTML이 없거나 빈 문자열인 포스트 조회
// duration: 현재 시간으로부터 과거로 얼마나 떨어진 시점까지의 새로 추가된 데이터를 가져올지 결정하는 시간 간격
func (r *PostRepository) FindPostsWithoutRenderedHTML(ctx context.Context, limit int64, duration time.Duration) ([]models.Post, error) {
	targetTime := time.Now().Add(-duration)

	filter := bson.M{
		"$or": []bson.M{
			{"rendered_html": bson.M{"$exists": false}},
			{"rendered_html": ""},
		},
		"created_at": bson.M{"$lt": targetTime},
	}
	findOpts := options.Find().SetLimit(limit).SetSort(bson.D{
		{Key: "created_at", Value: 1},
		{Key: "_id", Value: 1},
	})

	cur, err := r.col.Find(ctx, filter, findOpts)
	if err != nil {
		return nil, err
	}
	defer cur.Close(ctx)

	var results []models.Post
	for cur.Next(ctx) {
		var p models.Post
		if err := cur.Decode(&p); err != nil {
			return nil, err
		}
		results = append(results, p)
	}
	if err := cur.Err(); err != nil {
		return nil, err
	}

	return results, nil
}

// FindPostsWithoutThumbnail ThumbnailURL이 없거나 빈 문자열이지만 RenderedHTML은 있는 포스트 조회
// duration: 현재 시간으로부터 과거로 얼마나 떨어진 시점까지의 새로 추가된 데이터를 가져올지 결정하는 시간 간격
func (r *PostRepository) FindPostsWithoutThumbnail(ctx context.Context, limit int64, duration time.Duration) ([]models.Post, error) {

	// 현재 시간에서 입력받은 duration을 뺀 시점을 계산합니다.
	targetTime := time.Now().Add(-duration)

	filter := bson.M{
		"$and": []bson.M{
			{
				"$or": []bson.M{
					{"thumbnail_url": bson.M{"$exists": false}},
					{"thumbnail_url": ""},
				},
			},
			{"rendered_html": bson.M{"$exists": true, "$ne": ""}},
			{"created_at": bson.M{"$lt": targetTime}},
		},
	}

	findOpts := options.Find().SetLimit(limit).SetSort(bson.D{
		{Key: "created_at", Value: 1},
		{Key: "_id", Value: 1},
	})

	cur, err := r.col.Find(ctx, filter, findOpts)
	if err != nil {
		return nil, err
	}
	defer cur.Close(ctx)

	var results []models.Post
	for cur.Next(ctx) {
		var p models.Post
		if err := cur.Decode(&p); err != nil {
			return nil, err
		}
		results = append(results, p)
	}
	if err := cur.Err(); err != nil {
		return nil, err
	}

	return results, nil
}

// FindPostsWithoutSummary AI 요약이 없지만 RenderedHTML은 있는 포스트 조회 (duration 인자 추가됨)
// duration: 현재 시간으로부터 과거로 얼마나 떨어진 시점까지의 새로 추가된 데이터를 가져올지 결정하는 시간 간격
func (r *PostRepository) FindPostsWithoutSummary(ctx context.Context, limit int64, duration time.Duration) ([]models.Post, error) {

	targetTime := time.Now().Add(-duration)

	filter := bson.M{
		"status.ai_summarized": false,
		"rendered_html":        bson.M{"$exists": true, "$ne": ""},
		"created_at":           bson.M{"$lt": targetTime},
	}

	findOpts := options.Find().SetLimit(limit).SetSort(bson.D{
		{Key: "created_at", Value: 1},
		{Key: "_id", Value: 1},
	})

	cur, err := r.col.Find(ctx, filter, findOpts)
	if err != nil {
		return nil, err
	}
	defer cur.Close(ctx)

	var results []models.Post
	for cur.Next(ctx) {
		var p models.Post
		if err := cur.Decode(&p); err != nil {
			return nil, err
		}
		results = append(results, p)
	}
	if err := cur.Err(); err != nil {
		return nil, err
	}

	return results, nil
}

File: content_service/app/repositories/post_repository.py
from __future__ import annotations

from datetime import datetime, timezone
import re
from typing import Iterable

from pymongo import IndexModel, ASCENDING, DESCENDING
from pymongo.database import Database

from .documents.post_document import PostDocument
from .interfaces import PostRepositoryInterface, TagCountRow, TagSeriesRow
from common.models.post import ListPostsFilter, Post
from common.mongo.types import from_object_id, to_object_id


class PostRepository(PostRepositoryInterface):
    """posts 컬렉션에 대한 MongoDB 접근 레이어.

    Go `repositories.PostRepository` 와 동일한 책임을 갖는다.
    """

    def __init__(self, database: Database) -> None:
        """Mongo Database를 의존성으로 받고, posts 컬렉션을 내부에서 선택한다."""

        self._db = database
        self._col = database["posts"]
        self._col.create_indexes(
            [
                IndexModel(
                    [("published_at", DESCENDING), ("_id", DESCENDING)],
                    name="idx_published_at_id_desc",
                ),
                IndexModel(
                    [("aisummary.categories", ASCENDING)], name="idx_categories"
                ),
                IndexModel([("aisummary.tags", ASCENDING)], name="idx_tags"),
                IndexModel(
                    [("aisummary.tags", ASCENDING), ("published_at", DESCENDING)],
                    name="idx_tags_published_at",
                ),
                IndexModel(
                    [("aisummary.categories", ASCENDING), ("published_at", DESCENDING)],
                    name="idx_categories_published_at",
                ),
                IndexModel([("link", ASCENDING)], name="uniq_link", unique=True),
            ]
        )

    # --- helpers -----------------------------------------------------------------
    @staticmethod
    def _to_document(post: Post) -> dict:
        doc = PostDocument.from_domain(post)

        # BaseDocument.to_mongo_record() 를 통해 Mongo-safe 직렬화를 일관되게 사용한다.
        return doc.to_mongo_record()

    @staticmethod
    def _from_document(doc: dict) -> Post:
        document = PostDocument.model_validate(doc)

        return Post(
            id=from_object_id(document.id),
            created_at=document.created_at,
            updated_at=document.updated_at,
            status=document.status,
            view_count=document.view_count,
            blog_id=str(document.blog_id),
            blog_name=document.blog_name,
            title=document.title,
            link=document.link,
            published_at=document.published_at,
            thumbnail_url=document.thumbnail_url,
            aisummary=document.aisummary,
            embedding=document.embedding,
        )

    # --- commands ----------------------------------------------------------------
    def is_exist_by_link(self, link: str) -> bool:
        """링크로 포스트 존재 여부를 확인한다."""

        found = self._col.find_one({"link": link}, {"_id": 1})
        return found is not None

    def insert(self, post: Post) -> str:
        """새 포스트를 삽입하고 생성된 ID 를 반환한다."""

        now = datetime.now(timezone.utc)
        post.updated_at = now

        doc = self._to_document(post)
        result = self._col.insert_one(doc)
        inserted_id = result.inserted_id
        if inserted_id is None:
            raise RuntimeError("insert failed: inserted_id is None (possible null _id)")
        return str(inserted_id)

    def find_by_link(self, link: str) -> Post | None:
        doc = self._col.find_one({"link": link})
        if not doc:
            return None
        return self._from_document(doc)

    def list(self, flt: ListPostsFilter) -> tuple[list[Post], int]:
        """필터/페이지네이션 기준으로 포스트 목록과 총 개수를 반환한다."""

        filter_doc: dict = {}

        def _to_regex_in(values: Iterable[str]) -> list[re.Pattern]:
            arr: list[re.Pattern] = []
            for v in values:
                v = v.strip()
                if not v:
                    continue
                arr.append(re.compile(f"^{v}$", re.IGNORECASE))
            return arr

        cats_regex = _to_regex_in(flt.categories)
        tags_regex = _to_regex_in(flt.tags)

        if cats_regex and tags_regex:
            filter_doc["$or"] = [
                {"aisummary.categories": {"$in": cats_regex}},
                {"aisummary.tags": {"$in": tags_regex}},
            ]
        elif cats_regex:
            filter_doc["aisummary.categories"] = {"$in": cats_regex}
        elif tags_regex:
            filter_doc["aisummary.tags"] = {"$in": tags_regex}

        if flt.blog_id:
            filter_doc["blog_id"] = to_object_id(flt.blog_id)
        if flt.blog_name:
            filter_doc["blog_name"] = {"$regex": f"^{flt.blog_name}$", "$options": "i"}
        if flt.published_from or flt.published_to:
            published_range: dict[str, datetime] = {}
            if flt.published_from:
                published_range["$gte"] = flt.published_from
            if flt.published_to:
                published_range["$lte"] = flt.published_to
            filter_doc["published_at"] = published_range

        # Status 필터링: false 조회 시 필드가 없는 경우도 포함
        if flt.status_ai_summarized is not None:
            if flt.status_ai_summarized:
                filter_doc["status.ai_summarized"] = True
            else:
                # false이거나 필드가 없는 경우 모두 매칭
                filter_doc["$and"] = filter_doc.get("$and", []) + [
                    {
                        "$or": [
                            {"status.ai_summarized": False},
                            {"status.ai_summarized": {"$exists": False}},
                        ]
                    }
                ]
        if flt.status_embedded is not None:
            if flt.status_embedded:
                filter_doc["status.embedded"] = True
            else:
                filter_doc["$and"] = filter_doc.get("$and", []) + [
                    {
                        "$or": [
                            {"status.embedded": False},
                            {"status.embedded": {"$exists": False}},
                        ]
                    }
                ]

        page = flt.page if flt.page > 0 else 1
        page_size = flt.page_size
        if page_size <= 0 or page_size > 100:
            page_size = 20

        skip = (page - 1) * page_size

        total = self._col.count_documents(filter_doc)

        cursor = self._col.find(
            filter_doc,
            {"plain_text": 0},
            sort=[("published_at", -1), ("_id", -1)],
            skip=skip,
            limit=page_size,
        )

        items: list[Post] = []
        for doc in cursor:
            items.append(self._from_document(doc))

        return items, total

    def list_by_ids(self, ids: list[str]) -> list[Post]:
        """지정된 ObjectID 목록에 해당하는 포스트들을 반환한다."""

        if not ids:
            return []

        object_ids = [to_object_id(v) for v in ids]

        cursor = self._col.find(
            {"_id": {"$in": object_ids}},
            {"plain_text": 0},
        )

        items: list[Post] = []
        for doc in cursor:
            items.append(self._from_document(doc))

        return items

    def find_by_id(self, id_value: str) -> Post | None:
        doc = self._col.find_one(
            {"_id": to_object_id(id_value)},
            {"plain_text": 0},
        )
        if not doc:
            return None
        return self._from_document(doc)

    def get_plain_text(self, id_value: str) -> str | None:
        doc = self._col.find_one(
            {"_id": to_object_id(id_value)},
            {"plain_text": 1},
        )
        if not doc:
            return None
        value = doc.get("plain_text")
        if value is None:
            return ""
        return str(value)

    def increment_view_count(self, id_value: str) -> bool:
        result = self._col.update_one(
            {"_id": to_object_id(id_value)},
            {
                "$inc": {"view_count": 1},
                "$set": {"updated_at": datetime.now(timezone.utc)},
            },
        )
        return result.matched_count > 0

    def update_fields(self, id_value: str, updates: dict) -> None:
        allowed_keys = {
            "plain_text",
            "thumbnail_url",
            "aisummary",
            "status",
            "embedding",
        }
        invalid_keys = [key for key in updates.keys() if key not in allowed_keys]
        if invalid_keys:
            raise ValueError(f"unsupported update fields: {invalid_keys}")

        set_doc = {"updated_at": datetime.now(timezone.utc)}
        set_doc.update(updates)
        self._col.update_one(
            {"_id": to_object_id(id_value)},
            {"$set": set_doc},
        )

    def delete_by_id(self, id_value: str) -> bool:
        result = self._col.delete_one({"_id": to_object_id(id_value)})
        return result.deleted_count > 0

    def delete_by_blog_id(self, blog_id: str) -> int:
        result = self._col.delete_many({"blog_id": to_object_id(blog_id)})
        return int(result.deleted_count)

    def list_ids_by_blog_id(self, blog_id: str) -> list[str]:
        cursor = self._col.find({"blog_id": to_object_id(blog_id)}, {"_id": 1})
        return [
            id_value
            for doc in cursor
            if (id_value := from_object_id(doc.get("_id"))) is not None
        ]

    def count_by_blog_ids(self, blog_ids: list[str]) -> dict[str, int]:
        if not blog_ids:
            return {}

        object_ids = [to_object_id(blog_id) for blog_id in blog_ids]
        pipeline = [
            {"$match": {"blog_id": {"$in": object_ids}}},
            {"$group": {"_id": "$blog_id", "count": {"$sum": 1}}},
        ]

        counts: dict[str, int] = {}
        for doc in self._col.aggregate(pipeline):
            blog_id = from_object_id(doc["_id"])
            if blog_id is not None:
                counts[blog_id] = int(doc["count"])
        return counts

    def get_tag_counts_between(
        self, published_from: datetime, published_to: datetime
    ) -> list[TagCountRow]:
        pipeline = [
            {
                "$match": {
                    "published_at": {"$gte": published_from, "$lt": published_to},
                    "status.ai_summarized": True,
                    "aisummary.tags": {"$exists": True, "$type": "array", "$ne": []},
                }
            },
            {"$unwind": "$aisummary.tags"},
            {"$match": {"aisummary.tags": {"$type": "string", "$ne": ""}}},
            {
                "$group": {
                    "_id": {"$toLower": "$aisummary.tags"},
                    "original": {"$first": "$aisummary.tags"},
                    "count": {"$sum": 1},
                }
            },
            {"$sort": {"count": -1, "original": 1}},
        ]

        rows: list[TagCountRow] = []
        for doc in self._col.aggregate(pipeline):
            key = str(doc["_id"]).strip()
            if not key:
                continue
            rows.append(
                {
                    "key": key,
                    "tag": str(doc.get("original") or key),
                    "count": int(doc["count"]),
                }
            )
        return rows

    def get_tag_series(
        self,
        tags: list[str],
        published_from: datetime,
        published_to: datetime,
        interval: str,
    ) -> list[TagSeriesRow]:
        if interval not in {"day", "week", "month"}:
            raise ValueError(f"unsupported trend interval: {interval}")

        tag_patterns = [
            re.compile(f"^{re.escape(tag.strip())}$", re.IGNORECASE)
            for tag in tags
            if tag.strip()
        ]
        if not tag_patterns:
            return []

        match_doc = {
            "published_at": {"$gte": published_from, "$lt": published_to},
            "status.ai_summarized": True,
            "aisummary.tags": {"$in": tag_patterns},
        }

        pipeline = [
            {"$match": match_doc},
            {"$unwind": "$aisummary.tags"},
            {"$match": {"aisummary.tags": {"$in": tag_patterns}}},
            {
                "$group": {
                    "_id": {
                        "tag": {"$toLower": "$aisummary.tags"},
                        "bucket": {
                            "$dateTrunc": {
                                "date": "$published_at",
                                "unit": interval,
                                "timezone": "UTC",
                            }
                        },
                    },
                    "original": {"$first": "$aisummary.tags"},
                    "post_count": {"$sum": 1},
                    "blog_ids": {"$addToSet": "$blog_id"},
                }
            },
            {
                "$project": {
                    "_id": 0,
                    "key": "$_id.tag",
                    "tag": "$original",
                    "bucket": "$_id.bucket",
                    "post_count": 1,
                    "blog_count": {"$size": "$blog_ids"},
                }
            },
            {"$sort": {"bucket": 1, "tag": 1}},
        ]

        rows: list[TagSeriesRow] = []
        for doc in self._col.aggregate(pipeline):
            rows.append(
                {
                    "key": str(doc["key"]),
                    "tag": str(doc["tag"]),
                    "bucket": doc["bucket"],
                    "post_count": int(doc["post_count"]),
                    "blog_count": int(doc["blog_count"]),
                }
            )
        return rows

    def get_category_stats(
        self, blog_id: str | None, tags: list[str]
    ) -> dict[str, int]:
        """카테고리별 포스트 개수를 반환한다. (대소문자 무시)"""
        filter_doc: dict = {}

        if blog_id:
            filter_doc["blog_id"] = to_object_id(blog_id)

        if tags:
            tags_regex = [
                re.compile(f"^{tag.strip()}$", re.IGNORECASE)
                for tag in tags
                if tag.strip()
            ]
            if tags_regex:
                filter_doc["aisummary.tags"] = {"$in": tags_regex}

        pipeline = [
            {"$match": filter_doc},
            {"$unwind": "$aisummary.categories"},
            {
                "$group": {
                    "_id": {"$toLower": "$aisummary.categories"},
                    "original": {"$first": "$aisummary.categories"},
                    "count": {"$sum": 1},
                }
            },
        ]

        result = {}
        for doc in self._col.aggregate(pipeline):
            result[doc["original"]] = doc["count"]

        return result

    def get_tag_stats(
        self, blog_id: str | None, categories: list[str]
    ) -> dict[str, int]:
        """태그별 포스트 개수를 반환한다. (대소문자 무시)"""
        filter_doc: dict = {}

        if blog_id:
            filter_doc["blog_id"] = to_object_id(blog_id)

        if categories:
            cats_regex = [
                re.compile(f"^{cat.strip()}$", re.IGNORECASE)
                for cat in categories
                if cat.strip()
            ]
            if cats_regex:
                filter_doc["aisummary.categories"] = {"$in": cats_regex}

        pipeline = [
            {"$match": filter_doc},
            {"$unwind": "$aisummary.tags"},
            {
                "$group": {
                    "_id": {"$toLower": "$aisummary.tags"},
                    "original": {"$first": "$aisummary.tags"},
                    "count": {"$sum": 1},
                }
            },
        ]

        result = {}
        for doc in self._col.aggregate(pipeline):
            result[doc["original"]] = doc["count"]

        return result

    def get_blog_stats(
        self, categories: list[str], tags: list[str]
    ) -> list[tuple[str, str, int]]:
        """블로그별 포스트 개수를 반환한다. (blog_id, blog_name, count)"""
        filter_doc: dict = {}

        cats_regex = []
        tags_regex = []

        if categories:
            cats_regex = [
                re.compile(f"^{cat.strip()}$", re.IGNORECASE)
                for cat in categories
                if cat.strip()
            ]

        if tags:
            tags_regex = [
                re.compile(f"^{tag.strip()}$", re.IGNORECASE)
                for tag in tags
                if tag.strip()
            ]

        if cats_regex and tags_regex:
            filter_doc["$or"] = [
                {"aisummary.categories": {"$in": cats_regex}},
                {"aisummary.tags": {"$in": tags_regex}},
            ]
        elif cats_regex:
            filter_doc["aisummary.categories"] = {"$in": cats_regex}
        elif tags_regex:
            filter_doc["aisummary.tags"] = {"$in": tags_regex}

        pipeline = [
            {"$match": filter_doc},
            {
                "$group": {
                    "_id": "$blog_id",
                    "blog_name": {"$first": "$blog_name"},
                    "count": {"$sum": 1},
                }
            },
        ]

        result = []
        for doc in self._col.aggregate(pipeline):
            blog_id = from_object_id(doc["_id"])
            blog_name = doc["blog_name"]
            count = doc["count"]
            result.append((blog_id, blog_name, count))

        return result
